conn.go

```diff
@@ -172,7 +172,7 @@ func Connect(addr string, cfg *ConnConfig, errorHandler ConnErrorHandler, sessio
 		r:            bufio.NewReader(conn),
 		cfg:          cfg,
 		uniq:         make(chan int, streams),
-		calls:        make([]callReq, streams),
+		calls:        make([]*callReq, streams), // TODO(zariel): use a sparse map instead?
```

Comment:
IMO, i'd be great if you could get rid of allocating `streams` upfront entirely. The map suggestion is apt, but the lock might be undesirable. Motivation being is this is allocated for every connection + combined with something like DiscoverHosts, and 2 hosts per connection, large clusters tend to require a lot of idle memory. Tracking the current inflight streams seems to be the deal breaker - however, if you could use the `calls` data structure to enforce `cfg.NumStreams` (possibly blocking once we reach `cfg.NumStreams`, you could turn `uniq` into an atomic counter.